rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+function, without any significant syntactic complexity.
+
+### To Bang Or Not To Bang
+
+The proposal is to introduce the concept of the bang (`!`) function identifier
+additive to determine as a user if a function should be allowed to fail or not.
+
+Function identifiers with a bang are considered to be fallible, while functions
+without are not:
+
+```javascript
+.message = parse_json(.message) // cannot fail
+.counter = 1 + to_int!(.counter) // allowed to fail
+.timestamp = now()
+```
+
+In the above example, `parse_json` is written without a bang, and will thus
+never fail, whereas `to_int!` includes the bang, and will fail the program (and
+leave the event unchanged) if `.counter` cannot be coerced into an integer.
+
+### Infallible Function Result
+
+So far so good, but telling a program you don't want JSON parsing the fail does
+not magically make any input valid JSON.
+
+Given the original example, any badly formatted input values are still
+considered invalid JSON and the parser is thus unable to return a representative
+JSON value.
+
+So what should the function return instead?
+
+Infallible functions are allowed to return `Option<Value>`. In other words, they
+can return "nothing", if there isn't anything useful to return.
+
+### How To Handle "Nothing"
+
+Let's take a look at the original example again:
+
+```javascript
+.message = parse_json(.message)
+.timestamp = now()
+```
+
+If `message` contains any data that `parse_json` can't handle, it will return
+`None`.
+
+How that `None` gets processed is up to the expression that processes the value.
+
+For example, in the case of the `Assignment` expression (which is what's called
+when doing `.message = …`), `None` values skip assigning the value to the target
+path, returning `None` itself.
+
+Given this, the outcome of the above example would change from "exception,
+original event untouched", to "JSON parsing failed, `message` field untouched,
+`timestamp` field set to current timestamp".
+
+### Explicitly Handle Error Case
+
+While the above provides more flexibility to the end-user on when to terminate
+the program and when not to, it still does not allow the user to detect if JSON
+parsing failed.
+
+To do this, we modify the `IfStatement` and `Logical` expressions to support
+falling though to the "else" expression if the conditional expression returns
+`None`.
+
+For example:
+
+```javascript
+if ($message = parse_json(.message)) {
+    // parsing succeeded
+} else {
+    // parsing failed
+}
+```
+
+### False vs. "Nothing"
+
+One issue with this proposal is that `false` is a valid JSON value, and so the
+above check does not distinguish between false values or failed JSON parsing.
+
+One solution to this is to _not_ update `IfStatement`, but instead introduce a
+new `is_undefined` function, which returns `true` for `None` values, or else
+returns `false` (or inversely, we add a `is_defined` function):
```

Comment:
If we tweak this a bit, you could imagine an `ok(expr)` function that is essentially `expr != None`. If assignment returns the assigned value, you could have something like:

```
if ok($message = parse_json(.message)) {
    ...
} else {
    ...
}
```

Alternatively, we could have assignment return a boolean of whether or not a value was assigned and skip the extra function altogether for this case.